libraries/objectDefaultFiles/glCommandBuffer.js

```diff
@@ -0,0 +1,1073 @@
+                    commandBufferId: this.commandBufferId,
+                    isRendering: false,
+                    commands: this.commands,
+                });
+                postMessage({
+                    workerId: this.workerId,
+                    isFrameEnd: true
+                });
+                // await response
+                Atomics.wait(this.synclock, 0, 0);
+            }
+            catch (e) {
+                console.error(e);
+            }
+        }
+
+    }
+
+    // clears the whole buffer for reuse
+    clear() {
+        this.commands = [];
+        this.isCleared = true; // set to indicate that the buffer has been cleared since last time this flag was set to false
+    }
+}
+
+/**
+ * creates new command buffer using the hidden variables to hide construction details
+ * 
+ */
+class CommandBufferFactory {
+    /**
+     * 
+     * @param {number} workerId this worker's id
+     * @param {GLCommandBufferContext} glCommandBufferContext the context used by outr fake webgl to find to correct commandbuffer to write to
+     * @param {Int32Array} synclock synchronisation mechnism for synchroneous commands
+     */
+    constructor(workerId, glCommandBufferContext, synclock) {
+        this.workerId = workerId;
+        this.synclock = synclock;
+        this.glCommandBufferContext = glCommandBufferContext;
+    }
+
+    /**
+     * 
+     * @returns the fake webgl renderer used by this commandbuffer factory
+     */
+    getGL() {
+        return this.glCommandBufferContext.gl;
+    }
+
+    /**
+     * creates a new command buffer and immediately activates it. new commands executed on the fake webgl context will be stored in the newly created buffer
+     * @param {boolean} isRendering indicates that the commands stored in this buffer are meant to be repeatable
+     * @returns {CommandBuffer} th newly created and activated commandbuffer
+     */
+    createAndActivate(isRendering) {
+        let commandBuffer = new CommandBuffer(this.workerId, this.synclock, isRendering);
+        this.glCommandBufferContext.setActiveCommandBuffer(commandBuffer);
+        return commandBuffer;
+    }
+}
+
+/**
+ * manages all received command buffers on the server
+ * this class will decide when to execute a resource command buffer and which render command buffer to use for rendering the proxy
+ */
+class CommandBufferManager {
+    /**
+     * 
+     * @param {number} workerId 
+     */
+    constructor(workerId) {
+        /**
+         * all received command buffers in order
+         * @type {CommandBuffer[]}
+         */
+        this.commandBuffers = [];
+
+        /**
+         * the commandbuffer currently used for repeated rendering
+         * the rendering buffer starts as an empty commandbuffer, so the tool doesn't show until it is fully loaded.
+         * @type {CommandBuffer}
+         */
+        this.renderCommandBuffer = new CommandBuffer(workerId, -1, true, []);
```

Comment:
What is the effect of initializing a commandbuffer with synclock set to -1 here? It looks like this is a placeholder so `null` would be a bit clearer in purpose. It also looks like CommandBuffer doesn't handle an optional synclock

Reply:
I don't know why I made it -1, null indeed is much better, this code is server side only and makes sure there is always an empty render buffer to render from. This buffer doesn't need sync message support because it is never send.